graph/src/blockchain/firehose_block_ingestor.rs

```diff
@@ -0,0 +1,156 @@
+use std::{marker::PhantomData, sync::Arc, time::Duration};
+
+use crate::{
+    blockchain::Block as BlockchainBlock,
+    components::store::ChainStore,
+    firehose::{bstream, decode_firehose_block, endpoints::FirehoseEndpoint},
+    prelude::{error, info, Logger},
+    util::backoff::ExponentialBackoff,
+};
+use anyhow::{Context, Error};
+use futures03::StreamExt;
+use slog::trace;
+use tonic::Streaming;
+pub struct FirehoseBlockIngestor<M>
+where
+    M: prost::Message + BlockchainBlock + Default + 'static,
+{
+    chain_store: Arc<dyn ChainStore>,
+    endpoint: Arc<FirehoseEndpoint>,
+    logger: Logger,
+
+    phantom: PhantomData<M>,
+}
+
+impl<M> FirehoseBlockIngestor<M>
+where
+    M: prost::Message + BlockchainBlock + Default + 'static,
+{
+    pub fn new(
+        chain_store: Arc<dyn ChainStore>,
+        endpoint: Arc<FirehoseEndpoint>,
+        logger: Logger,
+    ) -> FirehoseBlockIngestor<M> {
+        FirehoseBlockIngestor {
+            chain_store,
+            endpoint,
+            logger,
+            phantom: PhantomData {},
+        }
+    }
+
+    pub async fn run(self) {
+        let mut latest_cursor = self.fetch_head_cursor().await;
+        let mut backoff =
+            ExponentialBackoff::new(Duration::from_millis(250), Duration::from_secs(30));
+
+        loop {
+            info!(
+                self.logger,
+                "Blockstream disconnected, connecting"; "endpoint uri" => format_args!("{}", self.endpoint), "cursor" => format_args!("{}", latest_cursor),
+            );
+
+            let result = self
+                .endpoint
+                .clone()
+                .stream_blocks(bstream::BlocksRequestV2 {
+                    start_block_num: -1,
```

Comment:
Will add, note also the hovering the field(s) give some extra documentation (picked from the protobuf source files when they are generated).

For the cursor vs start block, the reasoning to make it easier to code instead of having two code path:

```
if cursor != <empty>
  StartStream(..., cursor: ...)
else
  StartStream(start_block, ...)
end
```